gps: add status and integrity information

flyingthingsintothings opened the pull request:
Add fields to report resilience and status information from GNSS receivers back to ground control stations for representation in the user interface.

flyingthingsintothings commented:
This is a proposal to add some status information to the MAVLink protocol for GNSS receivers to report information like jamming, spoofing, errors... back to ground control stations, so they can report them to the user. It's part of feature proposals to PX4, ArduPilot, QGroundControl and Mission Planner where all of this is implemented based on this change.

peterbarker commented:
I wonder if this data should be in a separate message.

You haven't mentioned it here, but these same changes would be required in `GPS2_RAW`.

I've been mulling over a new message to replace these RAW messages as ArduPilot will soon have support for >2 receivers.

flyingthingsintothings commented:
I thought about that as well. The added resilience information would not require the same update rate as the other GPS data. Sending it every 1, 2 or even 5 seconds could also be fine, which isn't possible when it's bundled with the other data. I can add the changes to GPS2_RAW but that might not be the best approach when the messages will have to be changed anyway.

Should I already make a proposal for a new message containing this information? Would a new one use IDs for each receiver instead of different message types? Maybe this can also be discussed during the MAVLink community call this week?

hamishwillee commented:
@peterbarker What were you thinking about in terms of the new GPS_RAW message?

A few thoughts:
1. Jamming and spoofing likely to be common to all the GPS systems right, so maybe they might be in https://mavlink.io/en/messages/common.html#GPS_STATUS
2. I like the idea of an instance number and a new message.
3. Similar pattern used for https://mavlink.io/en/messages/common.html#GPS_RTK - are we potentially having more than 2 of those
4. We could make it GNSS_RAW
5. This would be a good time to think about our altitude elipsoid reference
6. Are there any other fields we've felt are missing for a while?

@flyingthingsintothings We'll probably raise this in the meeting, but I don't think we'll resolve anything this week.

tridge commented on message_definitions/v1.0/common.xml:
```diff
@@ -5079,6 +5079,78 @@
+      <entry value="2" name="GPS_SPOOFING_STATE_DETECTED">
+        <description>The GPS receiver detected signal spoofing.</description>
+      </entry>
+      <entry value="3" name="GPS_SPOOFING_STATE_MITIGATED">
```
i'd make a higher number worse, with the DETECTED be STATE_CRITICAL, meaning detected and fix is compromised
ref ublox:
![image](https://github.com/mavlink/mavlink/assets/831867/fda858c6-2397-47d9-86b8-3caf43147f15)

tridge reviewed (commented):
should be able to encode the ublox values as well:
![image](https://github.com/mavlink/mavlink/assets/831867/78b2d184-8bd2-4d40-b6c6-2ab8ca6afefd)

tridge commented:
@hamishwillee a new instanced msg does sound like a good idea, especially as we add support for 3 or more GPS receivers on an aircraft

tridge commented:
I suggest we add a GNSS_INTEGRITY message which is instanced, and it would only be sent if the GPS is capable of producing the information
I think it should also include the needed information to implement RAIM
see for example this RAIM msg in uAvionix:
![image](https://github.com/mavlink/mavlink/assets/831867/9492010a-aeef-43d3-9fab-bbf3265e3ae2)

julianoes commented on message_definitions/v1.0/common.xml:
```diff
@@ -5079,6 +5079,78 @@
     </enum>
+    <enum name="GPS_SYSTEM_ERROR_FLAGS" bitmask="true">
+      <description>Flags indicating errors in a GPS receiver.</description>
+      <entry value="0" name="GPS_SYSTEM_ERROR_NONE">
+        <description>There are no errors in the GPS receiver.</description>
+      </entry>
```
If these are bitflags, we probably don't need this one.

julianoes reviewed (approved):
Just one small comment, otherwise this is fine with me.

flyingthingsintothings commented:
@tridge I looked into RAIM and what would be needed to support it. In the uAvionix document that you shared a screenshot from, there are a lot of fields containing information about this. In other documentation (for u-blox and Septentrio for example), I can only find limited output that contains horizontal and vertical position certainty. I pushed a commit that adds RAIM state and position certainty, but I don't know whether it's enough or correct...

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -7686,6 +7773,17 @@
     </message>
+    <message id="441" name="GNSS_INTEGRITY">
+      <description>Information about key components of GNSS receivers, like signal authentication, interference and system errors.</description>
+      <field type="uint8_t" name="instance">Identifier for the GNSS receiver</field>
```
```suggestion
      <field type="uint8_t" name="id" instance="true">GNSS receiver id. Must match instance ids of other messages from same receiver.</field>
```

hamishwillee commented:
@tridge Any comments?

FYI @flyingthingsintothings Generally new messages go in development.xml until they have an implementation, and commitment from flight stacks to take them.

flyingthingsintothings commented:
Oh I didn't know about that, thanks for letting me know. I'll quickly move the message under there. The reason I added it to `common.xml` was because I already implemented some PRs with it in the autopilot projects and they use only `ardupilotmega.xml`, `common.xml` and `standard.xml` if I remember correctly. I will look into how to use `development.xml` for those PRs if that's the usual approach.

hamishwillee commented:
Thank you @flyingthingsintothings - appreciate you moving the message. We're trying to find better ways to stop things that are untested from getting into common.xml until they are ready - if we have to we use `<wip>` tags in common, but development.xml is used for new messages.

@tridge Is what @flyingthingsintothings did in https://github.com/mavlink/mavlink/pull/2110#issuecomment-2100230224 sufficient.  @peterbarker @auturgy Anything else needed?

SeptentrioGNSS commented:
@hamishwillee , @tridge , (@flyingthingsintothings) could the inclusion of 'mitigated' in the **Jamming** part of this message, the exclusion of 'critical' from **Spoofing** be considered for this message? 

Reasoning; aligning the jamming vs spoofing flags more, critical vs detected in the spoofing scenario would be identical.

flyingthingsintothings commented:
@SeptentrioGNSS Applied the changes.

tridge commented on message_definitions/v1.0/development.xml:
```diff
@@ -558,6 +645,17 @@
+      <field type="uint8_t" name="jamming_state" enum="GPS_JAMMING_STATE">Signal jamming state of the GPS system</field>
+      <field type="uint8_t" name="spoofing_state" enum="GPS_SPOOFING_STATE">Signal spoofing state of the GPS system</field>
+      <field type="uint8_t" name="raim_state" enum="GPS_RAIM_STATE">The state of the RAIM processing</field>
+      <field type="uint16_t" name="raim_hfom">Horizontal expected accuracy using satellites successfully validated using RAIM</field>
```
need units, units="cm"

tridge commented on message_definitions/v1.0/development.xml:
```diff
@@ -558,6 +645,17 @@
+      <field type="uint8_t" name="spoofing_state" enum="GPS_SPOOFING_STATE">Signal spoofing state of the GPS system</field>
+      <field type="uint8_t" name="raim_state" enum="GPS_RAIM_STATE">The state of the RAIM processing</field>
+      <field type="uint16_t" name="raim_hfom">Horizontal expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint16_t" name="raim_vfom">Vertical expected accuracy using satellites successfully validated using RAIM</field>
```
units

tridge reviewed (commented):
looks good apart from missing units, thanks!

tridge commented on message_definitions/v1.0/development.xml:
```diff
@@ -558,6 +645,21 @@
+      <field type="uint8_t" name="corrections_quality" minValue="0" maxValue="10">An abstract value representing the estimated quality of incoming corrections.</field>
+      <field type="uint8_t" name="system_status_summary" minValue="0" maxValue="10">An abstract value representing the overall status of the receiver.</field>
+      <field type="uint8_t" name="gnss_signal_quality" minValue="0" maxValue="10">An abstract value representing the quality of incoming GNSS signals.</field>
+      <field type="uint8_t" name="post_processing_quality" minValue="0" maxValue="10">An abstract value representing the estimated PPK quality.</field>
```
for these abstract value the commment needs to say what value should be sent if unknown. Also, can you explain how these values would be determined for a ublox F9?

flyingthingsintothings replied:
For receivers that do not yet provide specific information, such as the u-blox F9, the driver can generate this data based on other raw data provided and supported today by GNSS receivers. For instance, u-blox receivers output information about signal-to-noise ratio in their UBX-NAV-SAT messages. Those messages also include quality indicators for each signal, which the driver can process into a general status message and provide in the `gnss_signal_quality` field (e.g., a value compared to an ideal C/N of about 40 dB). That approach can also be applied to other values derived from existing messages. The main advantage of that method is that it simplifies complex advanced fields into a few quality parameters, making it easier for the user to understand. It's also better to process this data closer to the flight controller instead of in ground control station software as it results in a consistent experience for users across ground control station software (values provided by flight controller) and doesn't require sending raw GNSS data over MAVLink, using more bandwidth.

hamishwillee commented on message_definitions/v1.0/development.xml:
```diff
@@ -558,6 +645,21 @@
+      <field type="uint8_t" name="raim_state" enum="GPS_RAIM_STATE">The state of the RAIM processing</field>
+      <field type="uint16_t" name="raim_hfom" units="cm">Horizontal expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint16_t" name="raim_vfom" units="cm">Vertical expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="int8_t" name="corrections_quality" minValue="-1" maxValue="10">An abstract value representing the estimated quality of incoming corrections, or -1 if not available.</field>
+      <field type="int8_t" name="system_status_summary" minValue="-1" maxValue="10">An abstract value representing the overall status of the receiver, or -1 if not available.</field>
+      <field type="int8_t" name="gnss_signal_quality" minValue="-1" maxValue="10">An abstract value representing the quality of incoming GNSS signals, or -1 if not available.</field>
+      <field type="int8_t" name="post_processing_quality" minValue="-1" maxValue="10">An abstract value representing the estimated PPK quality, or -1 if not available.</field>
```
Why not uints and use UINT8_MAX for invalid value? We would normally do this like so:

```suggestion
      <field type="uint8_t" name="corrections_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated quality of incoming corrections, or UINT8_MAX if not available.</field>
      <field type="uint8_t" name="system_status_summary" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the overall status of the receiver, or UINT8_MAX if not available.</field>
      <field type="uint8_t" name="gnss_signal_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the quality of incoming GNSS signals, or UINT8_MAX if not available.</field>
      <field type="uint8_t" name="post_processing_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated PPK quality, or UINT8_MAX if not available.</field>
```

flyingthingsintothings replied:
I didn't know that there was a convention for invalid values. I'll quickly change it.

hamishwillee commented on message_definitions/v1.0/development.xml:
```diff
@@ -558,6 +645,21 @@
+      <field type="uint8_t" name="jamming_state" enum="GPS_JAMMING_STATE">Signal jamming state of the GPS system</field>
+      <field type="uint8_t" name="spoofing_state" enum="GPS_SPOOFING_STATE">Signal spoofing state of the GPS system</field>
+      <field type="uint8_t" name="raim_state" enum="GPS_RAIM_STATE">The state of the RAIM processing</field>
+      <field type="uint16_t" name="raim_hfom" units="cm">Horizontal expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint16_t" name="raim_vfom" units="cm">Vertical expected accuracy using satellites successfully validated using RAIM</field>
```
Will these always be supplied? I.e. do they need an invalid value?

flyingthingsintothings replied:
I don't think every receiver can supply RAIM information, so they would need an invalid value. I'll also use `UINT16_MAX` like you suggested for the other field.

hamishwillee commented:
@tridge Other than conflicts, anything you would like to add?

flyingthingsintothings commented:
Quickly resolved the conflicts.

tridge commented on message_definitions/v1.0/development.xml:
```diff
@@ -594,6 +681,21 @@
+      <field type="uint16_t" name="raim_hfom" units="cm" invalid="UINT16_MAX">Horizontal expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint16_t" name="raim_vfom" units="cm" invalid="UINT16_MAX">Vertical expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint8_t" name="corrections_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated quality of incoming corrections, or -1 if not available.</field>
+      <field type="uint8_t" name="system_status_summary" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the overall status of the receiver, or -1 if not available.</field>
```
-1 doesn't fit in uint8_t, I think it should say 255

hamishwillee replied:
Good catch. We normally say UINT8_MAX but 255 is more readable.

```suggestion
      <field type="uint8_t" name="system_status_summary" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the overall status of the receiver, or 255 if not available.</field>
```

tridge commented on message_definitions/v1.0/development.xml:
```diff
@@ -594,6 +681,21 @@
+      <field type="uint16_t" name="raim_vfom" units="cm" invalid="UINT16_MAX">Vertical expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint8_t" name="corrections_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated quality of incoming corrections, or -1 if not available.</field>
+      <field type="uint8_t" name="system_status_summary" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the overall status of the receiver, or -1 if not available.</field>
+      <field type="uint8_t" name="gnss_signal_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the quality of incoming GNSS signals, or -1 if not available.</field>
```
255

hamishwillee replied:
```suggestion
      <field type="uint8_t" name="gnss_signal_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the quality of incoming GNSS signals, or 255 if not available.</field>
```

tridge commented on message_definitions/v1.0/development.xml:
```diff
@@ -594,6 +681,21 @@
+      <field type="uint8_t" name="corrections_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated quality of incoming corrections, or -1 if not available.</field>
+      <field type="uint8_t" name="system_status_summary" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the overall status of the receiver, or -1 if not available.</field>
+      <field type="uint8_t" name="gnss_signal_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the quality of incoming GNSS signals, or -1 if not available.</field>
+      <field type="uint8_t" name="post_processing_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated PPK quality, or -1 if not available.</field>
```
255

hamishwillee replied:
```suggestion
      <field type="uint8_t" name="post_processing_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated PPK quality, or 255 if not available.</field>
```

hamishwillee commented on message_definitions/v1.0/development.xml:
```diff
@@ -594,6 +681,21 @@
+      <field type="uint8_t" name="raim_state" enum="GPS_RAIM_STATE">The state of the RAIM processing</field>
+      <field type="uint16_t" name="raim_hfom" units="cm" invalid="UINT16_MAX">Horizontal expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint16_t" name="raim_vfom" units="cm" invalid="UINT16_MAX">Vertical expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint8_t" name="corrections_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated quality of incoming corrections, or -1 if not available.</field>
```
```suggestion
      <field type="uint8_t" name="corrections_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated quality of incoming corrections, or 255 if not available.</field>
```

hamishwillee commented on message_definitions/v1.0/development.xml:
```diff
@@ -594,6 +681,21 @@
       </description>
       <field type="uint8_t" name="seq">Sequence number. The value iterates sequentially whenever AVAILABLE_MODES changes (e.g. support for a new mode is added/removed dynamically).</field>
     </message>
+    <message id="441" name="GNSS_INTEGRITY">
+      <description>Information about key components of GNSS receivers, like signal authentication, interference and system errors.</description>
+      <field type="uint8_t" name="id" instance="true">GNSS receiver id. Must match instance ids of other messages from same receiver.</field>
+      <field type="uint32_t" name="system_errors" enum="GPS_SYSTEM_ERROR_FLAGS" display="bitmask">Errors in the GPS system</field>
+      <field type="uint8_t" name="authentication_state" enum="GPS_AUTHENTICATION_STATE">Signal authentication state of the GPS system</field>
+      <field type="uint8_t" name="jamming_state" enum="GPS_JAMMING_STATE">Signal jamming state of the GPS system</field>
+      <field type="uint8_t" name="spoofing_state" enum="GPS_SPOOFING_STATE">Signal spoofing state of the GPS system</field>
+      <field type="uint8_t" name="raim_state" enum="GPS_RAIM_STATE">The state of the RAIM processing</field>
+      <field type="uint16_t" name="raim_hfom" units="cm" invalid="UINT16_MAX">Horizontal expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint16_t" name="raim_vfom" units="cm" invalid="UINT16_MAX">Vertical expected accuracy using satellites successfully validated using RAIM</field>
+      <field type="uint8_t" name="corrections_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated quality of incoming corrections, or 255 if not available.</field>
+      <field type="uint8_t" name="system_status_summary" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the overall status of the receiver, or 255 if not available.</field>
+      <field type="uint8_t" name="gnss_signal_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the quality of incoming GNSS signals, or 255 if not available.</field>
+      <field type="uint8_t" name="post_processing_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated PPK quality, or 255 if not available.</field>
+    </message>
```
consistency
```suggestion
    <message id="441" name="GNSS_INTEGRITY">
      <description>Information about key components of GNSS receivers, like signal authentication, interference and system errors.</description>
      <field type="uint8_t" name="id" instance="true">GNSS receiver id. Must match instance ids of other messages from same receiver.</field>
      <field type="uint32_t" name="system_errors" enum="GPS_SYSTEM_ERROR_FLAGS" display="bitmask">Errors in the GPS system.</field>
      <field type="uint8_t" name="authentication_state" enum="GPS_AUTHENTICATION_STATE">Signal authentication state of the GPS system.</field>
      <field type="uint8_t" name="jamming_state" enum="GPS_JAMMING_STATE">Signal jamming state of the GPS system.</field>
      <field type="uint8_t" name="spoofing_state" enum="GPS_SPOOFING_STATE">Signal spoofing state of the GPS system.</field>
      <field type="uint8_t" name="raim_state" enum="GPS_RAIM_STATE">The state of the RAIM processing.</field>
      <field type="uint16_t" name="raim_hfom" units="cm" invalid="UINT16_MAX">Horizontal expected accuracy using satellites successfully validated using RAIM.</field>
      <field type="uint16_t" name="raim_vfom" units="cm" invalid="UINT16_MAX">Vertical expected accuracy using satellites successfully validated using RAIM.</field>
      <field type="uint8_t" name="corrections_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated quality of incoming corrections, or 255 if not available.</field>
      <field type="uint8_t" name="system_status_summary" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the overall status of the receiver, or 255 if not available.</field>
      <field type="uint8_t" name="gnss_signal_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the quality of incoming GNSS signals, or 255 if not available.</field>
      <field type="uint8_t" name="post_processing_quality" minValue="0" maxValue="10" invalid="UINT8_MAX">An abstract value representing the estimated PPK quality, or 255 if not available.</field>
    </message>
```

hamishwillee reviewed (approved):
Merging into development as this has been approved (offline) by @tridge and by @julianoes 

Thanks @flyingthingsintothings !

flyingthingsintothings commented:
Thank you for all the help with this, both during the calls and here. Sorry for the few consistency problems. When you look at these things for too long you tend to miss these things 🙂 Now onto implementing this in the rest of the autopilot ecosystem.

peterbarker commented:
Oh, bother.  No TimeUS field?

hamishwillee commented:
@peterbarker If you need it, PR it in. This is still in development.xml and subject to change.

SeptentrioGNSS commented:
@peterbarker , what would this TimeUS field contain (found that it is usec since system boot)? But the content of this message would be appended with the usec time of GPS receiver? 
What would be the benefit of having this and not a GPS/GNSS timestamped value like TOW WNc (time of week and week number count)?

peterbarker commented:
On Wed, 28 Aug 2024, SepTen3o wrote:

> @peterbarker , what would this TimeUS field contain (found that it is usec since system boot)? But the content of this message would be appended with the usec time of GPS receiver?
> What would be the benefit of having this and not a GPS/GNSS timestamped value like TOW WNc (time of week and week number count)?

I mis-spoke there; I really should have said "time_boot_ms" - most 
messages we stream have it.  It should let you correlate the GNSS message 
with the GNSS_INTEGRITY message, for example.

SeptentrioGNSS commented:
Aah for that reason, speaking on behalf of FlyingThingsIntoThings, should not cause issues at a first glance but deeper understanding of his implementation is required for full confirmation. He is currently unavailable to comment on the matter.

And this is an overall boot time or specific to the GPS content in the message (e.g. boot time of GPS/GNSS receiver or autopilot boot time?)

peterbarker commented:
On Wed, 28 Aug 2024, SepTen3o wrote:

> And this is an overall boot time or specific to the GPS content in the message (e.g. boot time of GPS/GNSS receiver or autopilot boot time?)

Auto-pilot boot time.  Have a search through common.xml for "time_boot_ms"

Peter
